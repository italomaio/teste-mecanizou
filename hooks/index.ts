export { useAuth } from "./useAuth";
export { useCookies } from "./useCookies";
export { useProducts } from "./useProducts";
