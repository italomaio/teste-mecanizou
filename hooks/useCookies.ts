import { useState, useEffect } from "react";

function useCookies(cookieName: string) {
  const [value, setValue] = useState<string | null>(null);

  useEffect(() => {
    const getCookie = (name: string) => {
      const match = document.cookie.match(
        new RegExp("(^| )" + name + "=([^;]+)")
      );
      return match ? decodeURIComponent(match[2]) : null;
    };

    setValue(getCookie(cookieName));
  }, [cookieName]);

  const setCookie = (value: string, days: number) => {
    const expires = new Date();
    expires.setTime(expires.getTime() + days * 24 * 60 * 60 * 1000);
    document.cookie = `${cookieName}=${encodeURIComponent(
      value
    )};expires=${expires.toUTCString()};path=/`;
    setValue(value);
  };

  const deleteCookie = () => {
    document.cookie = `${cookieName}=;expires=Thu, 01 Jan 1970 00:00:00 UTC;path=/`;
    setValue(null);
  };

  return { value, setCookie, deleteCookie };
}

export { useCookies };
