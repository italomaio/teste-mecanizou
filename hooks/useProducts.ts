import { productsService } from "@/services/products/service";
import { ProductQuery, ProductState } from "@/types/modules/products";
import { useState, useCallback, useEffect } from "react";

const useProducts = () => {
  const [state, setState] = useState<ProductState>({
    products: [],
    loading: false,
    error: null,
    total: 0,
    skip: 0,
    limit: 10,
  });

  const fetchProducts = useCallback(async (query?: ProductQuery) => {
    setState((prev) => ({ ...prev, loading: true, error: null }));

    try {
      const response = await productsService.getAll(query);
      const errorMessage = response.error?.message || null;

      setState((prev) => ({
        ...prev,
        loading: false,
        error: errorMessage,
        products: response.data?.products || [],
        total: response.data?.total || 0,
        skip: response.data?.skip || 0,
        limit: response.data?.limit || 10,
      }));
    } catch (error) {
      setState((prev) => ({
        ...prev,
        loading: false,
        error: (error as Error).message || "Unexpected error",
      }));
    }
  }, []);

  useEffect(() => {
    fetchProducts();
  }, []);

  return { ...state, fetchProducts };
};

export { useProducts };
