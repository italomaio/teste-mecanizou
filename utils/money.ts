export const formatter = new Intl.NumberFormat("pt-BR", {
  style: "currency",
  currency: "BRL",
  minimumFractionDigits: 2,
});

export function calculateDiscountPrice(
  price: number,
  discountPercentage: number
) {
  const discountFactor = discountPercentage / 100;
  const newPrice = price * (1 - discountFactor);

  return {
    oldPrice: formatter.format(price),
    newPrice: formatter.format(newPrice),
  };
}
