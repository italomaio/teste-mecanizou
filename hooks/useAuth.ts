import { useCallback, useEffect, useState } from "react";
import { useCookies } from "@/hooks";
import { authService } from "@/services/auth";
import { UserProfile } from "@/types/modules/auth";
import { redirect } from "next/navigation";

interface AuthState {
  isAuthenticated: boolean;
  user: UserProfile | null;
}

export const useAuth = () => {
  const { value: token, deleteCookie } = useCookies("token");
  const [authState, setAuthState] = useState<AuthState>({
    isAuthenticated: !!token,
    user: null,
  });

  const logout = async () => {
    const result = await authService.logout();

    if (!result.error)
      setAuthState(() => {
        return { isAuthenticated: false, user: null };
      });

    redirect("/auth/login");
  };

  const fetchUserData = useCallback(async () => {
    const result = await authService.getProfile();

    if (!result.error)
      setAuthState((prev) => ({
        ...prev,
        user: result.data?.user as UserProfile,
      }));
  }, []);

  useEffect(() => {
    (() => {
      fetchUserData();
    })();
  }, [token]);

  return {
    ...authState,
    logout,
  };
};
