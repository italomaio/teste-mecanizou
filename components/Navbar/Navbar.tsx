"use client";

import React, { useState } from "react";
import { Button, Logo, UserDropdown } from "@/components";
import { ThemeToggle } from "@/components";
import { cn } from "@/utils/classes";
import { useAuth } from "@/hooks";
import { AlignJustify, LogOut, X } from "lucide-react";

const Navbar: React.FC = () => {
  const [isOpen, setIsOpen] = useState(false);
  const { user, logout } = useAuth();

  return (
    <nav
      className="border-b border-gray-300 dark:border-zinc-700 px-4 py-2 flex items-center justify-between"
      aria-label="Main navigation"
    >
      <Logo className="invert dark:invert-25" animate={false} />

      <ThemeToggle aria-label="Toggle theme" />

      <Button
        variant="ghost"
        aria-label={isOpen ? "Close menu" : "Open menu"}
        aria-expanded={isOpen}
        aria-controls="mobile-menu"
        className="md:hidden p-2 flex flex-col"
        onClick={() => setIsOpen((s) => !s)}
      >
        {!isOpen ? <AlignJustify /> : <X />}
      </Button>

      <div
        className="hidden md:flex items-center justify-center space-x-4"
        role="menu"
        aria-label="User menu"
      >
        <UserDropdown user={user} onLogout={() => logout()} />
      </div>

      <div
        id="mobile-menu"
        aria-label="Mobile menu"
        data-testid="mobile-menu"
        className={cn(
          "absolute flex flex-col top-full z-50 inset-x-0 bg-white dark:bg-zinc-800 md:hidden transition-transform duration-300 p-5 border-t border-gray-300 dark:border-t-zinc-700 shadow-2xl",
          !isOpen ? "translate-y-0" : "-translate-y-full"
        )}
      >
        <div
          className="flex flex-col p-4 space-y-2 justify-end h-full flex-1"
          role="menu"
          aria-label="Mobile user actions"
        >
          <Button
            onClick={logout}
            variant="outline"
            size="lg"
            aria-label="Logout by mobile menu"
          >
            <LogOut />
            Logout
          </Button>
        </div>
      </div>
    </nav>
  );
};

export default Navbar;
