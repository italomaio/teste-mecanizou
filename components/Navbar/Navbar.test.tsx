import React from "react";
import { render, screen, fireEvent } from "@/tests/utils";
import Navbar from "./Navbar";
import { useAuth } from "@/hooks";
import {
  ButtonProps,
  LabelProps,
  LogoProps,
  UserDropdownProps,
} from "@/components";

jest.mock("@/components", () => ({
  Button: ({ children, ...props }: ButtonProps) => (
    <button {...props}>{children}</button>
  ),
  Label: ({ children }: LabelProps) => <label>{children}</label>,
  Logo: ({ className }: LogoProps) => <div className={className}>Logo</div>,
  UserDropdown: ({ user, onLogout }: UserDropdownProps) => (
    <div data-testid="user-dropdown-mock">
      <span
        role="button"
        aria-label="User menu"
        data-testid="user-name-display"
      >
        {user?.email}
      </span>
      <button onClick={onLogout} aria-label="Logout dropdown">
        Logout
      </button>
    </div>
  ),
  ThemeToggle: () => <button>ThemeToggle</button>,
}));

describe("Navbar", () => {
  const mockLogout = jest.fn();

  beforeEach(() => {
    (useAuth as jest.Mock).mockReturnValue({
      user: { name: "Test User" },
      logout: mockLogout,
    });
  });

  it("Should render correctly", () => {
    const { container } = render(<Navbar />);
    expect(container).toBeInTheDocument();
  });

  it("Should render the theme toggle button", () => {
    render(<Navbar />);
    expect(screen.getByText("ThemeToggle")).toBeInTheDocument();
  });

  it("Should call logout when logout button is clicked in the dropdown", () => {
    render(<Navbar />);
    const logoutButtonInDropdown = screen.getByRole("button", {
      name: "Logout dropdown",
    });

    fireEvent.click(logoutButtonInDropdown);
    expect(mockLogout).toHaveBeenCalledTimes(1);
  });

  it("Should toggle the mobile menu when the menu button is clicked", () => {
    render(<Navbar />);

    const menuButton = screen.getByLabelText("Open menu");
    const mobileMenu = screen.getByLabelText("Mobile menu");

    expect(mobileMenu).toHaveClass("translate-y-0");
    expect(mobileMenu).not.toHaveClass("-translate-y-full");

    fireEvent.click(menuButton);

    expect(mobileMenu).not.toHaveClass("translate-y-0");
    expect(mobileMenu).toHaveClass("-translate-y-full");
  });

  it("Should call logout when logout button is clicked in the mobile menu", () => {
    render(<Navbar />);
    const menuButton = screen.getByLabelText("Open menu");

    fireEvent.click(menuButton);
    fireEvent.click(
      screen.getByRole("button", { name: /Logout by mobile menu/i })
    );

    expect(mockLogout).toHaveBeenCalledTimes(1);
  });

  it("Should match snapshot", () => {
    const { asFragment } = render(<Navbar />);
    expect(asFragment()).toMatchSnapshot();
  });
});
