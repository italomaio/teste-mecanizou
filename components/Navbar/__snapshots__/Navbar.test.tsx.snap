// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Navbar Should match snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="Main navigation"
    class="border-b border-gray-300 dark:border-zinc-700 px-4 py-2 flex items-center justify-between"
  >
    <div
      class="invert dark:invert-25"
    >
      Logo
    </div>
    <button>
      ThemeToggle
    </button>
    <button
      aria-controls="mobile-menu"
      aria-expanded="false"
      aria-label="Open menu"
      class="md:hidden p-2 flex flex-col"
      variant="ghost"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-align-justify"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 12h18"
        />
        <path
          d="M3 18h18"
        />
        <path
          d="M3 6h18"
        />
      </svg>
    </button>
    <div
      aria-label="User menu"
      class="hidden md:flex items-center justify-center space-x-4"
      role="menu"
    >
      <div
        data-testid="user-dropdown-mock"
      >
        <span
          aria-label="User menu"
          data-testid="user-name-display"
          role="button"
        />
        <button
          aria-label="Logout dropdown"
        >
          Logout
        </button>
      </div>
    </div>
    <div
      aria-label="Mobile menu"
      class="absolute flex flex-col top-full z-50 inset-x-0 bg-white dark:bg-zinc-800 md:hidden transition-transform duration-300 p-5 border-t border-gray-300 dark:border-t-zinc-700 shadow-2xl translate-y-0"
      data-testid="mobile-menu"
      id="mobile-menu"
    >
      <div
        aria-label="Mobile user actions"
        class="flex flex-col p-4 space-y-2 justify-end h-full flex-1"
        role="menu"
      >
        <button
          aria-label="Logout by mobile menu"
          variant="outline"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-log-out"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m16 17 5-5-5-5"
            />
            <path
              d="M21 12H9"
            />
            <path
              d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
            />
          </svg>
          Logout
        </button>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
